### ISSUE TITLE
Json serialization of Result

### ISSUE THREAD

**Issue by FrancoisM** [vkhorikov/CSharpFunctionalExtensions#41]
I can't return Result from a webapi as in case of success, Error throws an exception of type 'System.InvalidOperationException'. Why not just return string.empty?

**Comment by FrancoisM** [comment]
Here is what I ended up using:

```
public class ResultDto
    {
        public bool IsFailure { get; set; }
        public bool IsSuccess { get; set; }
        public string Error { get; set; }

        public ResultDto() { }

        public ResultDto(Result result)
        {
            IsFailure = result.IsFailure;
            IsSuccess = result.IsSuccess;
            Error = IsFailure ? result.Error : string.Empty;
        }
    }

    public class ResultDto<T> : ResultDto
    {
        public T Value { get; set; }

        public ResultDto(Result<T> result)
        {
            IsFailure = result.IsFailure;
            IsSuccess = result.IsSuccess;
            Error = IsFailure ? result.Error : string.Empty;
            Value = IsSuccess ? result.Value : default(T);
        }
    }
```

**Comment by jaibeales** [comment]
I agree with this, I'm having similar issues basically it's not Serializable.

**Comment by vkhorikov** [comment]
I'd recommend having a separate structure for the data returned by API. Something like this: https://github.com/vkhorikov/AnemicDomainModel/blob/master/After/src/Api/Utils/Envelope.cs

Result is not actually meant to be used as a data contract.

**Comment by LRFalk01** [comment]
I don't know if this helps you guys, but I wrote an IHttpActionResult that handles the Result object. It returns 400 on Result.IsFailure and 200 on Result.IsSuccess. The 400 error message is whatever Result.Error is. If it is a Result<T>, the Result.Value is sent as the body of the 200:

I have extension methods at the bottom so I can call them in my ApiController as follows:
return this.Result(result);

https://gist.github.com/d19a678d90659c3866a04c168abbae8a

**Comment by FrancoisM** [comment]
This in line with Envelope which I finally use in lieu of my previous
ResultDto.

Le jeu. 16 nov. 2017 à 20:43, Lucas Falk <notifications@github.com> a
écrit :

**Comment by jai-avalara** [comment]
Whilst I agreee with above comments re not returning this in the API, I would still say if you see ISerializable being implemented you would expect that class to be Serializable. I guess you could argue it does't have the [Serializable] attribute therefore doesn't support serialization, but really it's just a bit surprising it isn't.

By the way how I discovered this was not through trying to return this in an API as a DTO but by implementing some audit logging using Mediatr behaviours (since we use Mediatr with a CQS pattern). Since I didn't want to change the return types for this sort of thing I just configured JSON.net to ignore the exceptions for now, until such time we fork Result, refactor or perhaps even submit a PR.

**Comment by jaibeales** [comment]
Whilst I agreee with above comments re not returning this in the API, I would still say if you see ISerializable being implemented you would expect that class to be Serializable. I guess you could argue it does't have the [Serializable] attribute therefore doesn't support serialization, but really it's just a bit surprising it isn't.

By the way how I discovered this was not through trying to return this in an API as a DTO but by implementing some audit logging using Mediatr behaviours (since we use Mediatr with a CQS pattern). Since I didn't want to change the return types for this sort of thing I just configured JSON.net to ignore the exceptions for now, until such time we fork Result, refactor or perhaps even submit a PR.

**Comment by vkhorikov** [comment]
@jaibeales feel free to submit a PR. Although I personally don't use `Result` in a situation that require serialization, I don't see any harm in having this functionality either.

**Comment by CalvinDale** [comment]
Having a Newtonsoft Json.NET-compliant serializable `Result` would be quite useful with Azure's Durable Functions for passing results between activity functions within an orchestration. Perhaps the `ResultDto` wrapper suggested by @FrancoisM would work. I haven't tried it yet.

**Comment by stefano77it** [comment]
should be resolved with commit #118

**Comment by johanndev** [comment]
Maybe I'm a bit obtuse - but I cannot figure out how to use the serialization support in conjunction with `System.Text.Json`. The unit tests only demonstrate the usage with the `BinaryFormatter`.

E.g., this naive code fails with an `ResultSuccessException`, as it tries to serialize the error property:
```
var okResult = Result.Success("Success!");
var serializedOkResult = JsonConvert.SerializeObject(okResult);
```

Can somebody give me a hint?
